programs/gateway/src/lib.rs

```diff
@@ -80,6 +131,106 @@ pub mod gateway {
+        signature: [u8; 64],
+        recovery_id: u8,
+        message_hash: [u8; 32],
+        nonce: u64,
+    ) -> Result<()> {
+        let pda = &mut ctx.accounts.pda;
+
+        if nonce != pda.nonce {
+            msg!(
+                "Mismatch nonce: provided nonce = {}, expected nonce = {}",
+                nonce,
+                pda.nonce,
+            );
+            return err!(Errors::NonceMismatch);
+        }
+
+        let mut concatenated_buffer = Vec::new();
+        concatenated_buffer.extend_from_slice(b"ZETACHAIN");
+        concatenated_buffer.push(InstructionId::Execute as u8);
+        concatenated_buffer.extend_from_slice(&pda.chain_id.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&nonce.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&amount.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&ctx.accounts.destination_program.key().to_bytes());
+        require!(
+            message_hash == hash(&concatenated_buffer[..]).to_bytes(),
+            Errors::MessageHashMismatch
+        );
+
+        msg!("Computed message hash: {:?}", message_hash);
+
+        let address = recover_eth_address(&message_hash, recovery_id, &signature)?;
+        if address != pda.tss_address {
+            msg!("ECDSA signature error");
+            return err!(Errors::TSSAuthenticationFailed);
+        }
+
+        // NOTE: have to manually create Instruction, pack it and invoke since there is no crate for contract
+        // since any contract with on_call instruction can be called
+        let instruction_data = CallableInstruction::OnCall {
+            amount,
+            sender,
+            data,
+        }
+        .pack();
+
+        // account metas for remaining accounts
+        let account_metas: Vec<AccountMeta> = ctx
+            .remaining_accounts
+            .iter()
+            .map(|account_info| {
+                // Check if the account is writable, then create the appropriate AccountMeta
+                if account_info.is_writable {
+                    AccountMeta::new(*account_info.key, account_info.is_signer)
+                } else {
+                    AccountMeta::new_readonly(*account_info.key, account_info.is_signer)
+                }
+            })
+            .collect();
+
+        let ix = Instruction {
+            program_id: ctx.accounts.destination_program.key(),
+            accounts: account_metas,
+            data: instruction_data,
+        };
+
+        // withdraw to destination program pda
+        pda.sub_lamports(amount)?;
+        ctx.accounts.destination_program_pda.add_lamports(amount)?;
+
+        // invoke destination program on_call function
+        invoke(&ix, ctx.remaining_accounts)?;
+
+        msg!(
+            "Execute done: destination contract = {}, amount = {}, sender = {:?}",
+            amount,
+            ctx.accounts.destination_program.key(),
+            sender,
+        );
+
+        pda.nonce += 1;
```

Comment:
I see this operation is at different location for the withdraw functions, we might want to consolidate its placement for code cleaniness, I would suggest incremented the nonce right after it has been checked